Block: make header-only state immutable

schildbach opened the pull request:
Blocks can't shapeshift between regular (with transactions) and header-only any more. The `transactions` field is made final.

Add public `isHeaderOnly()` and use it.

Note: there is a difference between header-only, and an empty list of transactions. Headers do not contain the transaction count in its serialized representation.


This is draft because I want to make use of the new method in other classes – currently it's used only internally. Still I'd be interested in early feedback.

msgilligan reviewed (changes requested):
Nicely done! I have a few requests/suggestions. The only one that I feel is a "blocker" for this PR is having `writeTransactions` write zero bytes when it is a header-only block.

(I know it is a `private` and short-circuiting the call in `bitcoinSerializeToStream` is a minor optimization, but I think we should be consistent about properly handling null/empty items without using or requiring special case handling)

One related question: Is a non-header block with zero transactions automatically and always considered invalid? Or, in other words, would we want to deserialize a block with a VarInt.sizeOf(0) for transaction count following the header? It might be good to add some JavaDoc comments to these methods explaining this.

schildbach commented:
> One related question: Is a non-header block with zero transactions automatically and always considered invalid? Or, in other words, would we want to deserialize a block with a VarInt.sizeOf(0) for transaction count following the header? It might be good to add some JavaDoc comments to these methods explaining this.

It's invalid for the network, because blocks require at least a coinbase txn. However, I think within bitcoinj we should allow them because that's how we construct blocks currently. That may change, due to how change the way blocks will be built. Maybe just the "block builder" can have an empty list of transactions in future.

We could/should add a Precondition when deserializing, if not already present. I'll look into that.

schildbach commented on core/src/main/java/org/bitcoinj/core/Block.java:
```diff
@@ -312,7 +305,8 @@ private void writeTransactions(OutputStream stream) throws IOException {
     protected void bitcoinSerializeToStream(OutputStream stream) throws IOException {
         writeHeader(stream);
-        writeTransactions(stream);
+        if (!isHeaderOnly())
```
Actually it's intended that it writes a 0 size for an empty list of transaction.

schildbach replied:
In the comment at the very top you said "header-only" and here you say "no transactions". That's two different cases.

schildbach commented on core/src/main/java/org/bitcoinj/core/Block.java:
```diff
@@ -276,8 +276,7 @@ private static List<Transaction> genesisTransactions() {
         int size = HEADER_SIZE;
-        List<Transaction> transactions = getTransactions();
-        if (transactions != null) {
+        if (!isHeaderOnly()) {
```
I'll maybe extract `transactionsSize()` to a private method and use a ternary, because then it's simple enough to read.

But that's for after this refactor.

schildbach commented on core/src/main/java/org/bitcoinj/core/Block.java:
```diff
@@ -131,7 +131,7 @@ public enum VerifyFlag {
     // If null, it means this object holds only the headers.
     // For testing only
-    @Nullable List<Transaction> transactions;
+    @Nullable final List<Transaction> transactions;
```
There is already `getTransactions()` and I want to deprecate it for a non-nullable `transactions()`.

But I ran into problems with tests so I didn't include this change in this preview.

schildbach commented:
Rebased, and added the precondition to `writeTransactions()`.

schildbach commented:
I pushed a second commit on top that I think is a natural follow-up.

schildbach commented:
I rebased, and changed the recently added `requreNonNull()` to the new method.

I previously addressed one comment that is currently not replied to nor marked as resolved.

This is ready for a final review.
